File: extensions/vs/ClaudeCodeVS/Messages/Protocol.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace ClaudeCodeVS.Messages
{
    // =========================================================================
    // Context primitives — match extensions/vscode/src/protocol.ts L12-41
    // =========================================================================

    public class FileContext
    {
        [JsonProperty("path")] public string Path { get; set; }
        [JsonProperty("content")] public string Content { get; set; }
        [JsonProperty("language")] public string Language { get; set; }
    }

    public class Selection
    {
        [JsonProperty("path")] public string Path { get; set; }
        [JsonProperty("startLine")] public int StartLine { get; set; }
        [JsonProperty("startChar")] public int StartChar { get; set; }
        [JsonProperty("endLine")] public int EndLine { get; set; }
        [JsonProperty("endChar")] public int EndChar { get; set; }
        [JsonProperty("text")] public string Text { get; set; }
    }

    public class Diagnostic
    {
        [JsonProperty("path")] public string Path { get; set; }
        [JsonProperty("line")] public int Line { get; set; }
        [JsonProperty("column")] public int Column { get; set; }
        [JsonProperty("message")] public string Message { get; set; }
        [JsonProperty("severity")] public string Severity { get; set; } // "error" | "warning" | "info" | "hint"
    }

    // =========================================================================
    // Outgoing messages — IDE → Bun (match protocol.ts L47-158)
    // =========================================================================

    public class OutgoingMessage
    {
        [JsonProperty("type")] public string Type { get; set; }
    }

    public class UserMessage : OutgoingMessage
    {
        [JsonProperty("message")] public UserMessageBody Message { get; set; }
        [JsonProperty("parent_tool_use_id")] public string ParentToolUseId { get; set; }
    }

    public class UserMessageBody
    {
        [JsonProperty("role")] public string Role { get; set; } = "user";
        [JsonProperty("content")] public object Content { get; set; }
    }

    public class IdeContextMessage : OutgoingMessage
    {
        [JsonProperty("files")] public List<FileContext> Files { get; set; }
        [JsonProperty("selection")] public Selection Selection { get; set; }
        [JsonProperty("diagnostics")] public List<Diagnostic> Diagnostics { get; set; }
    }

    public class ControlResponseMessage : OutgoingMessage
    {
        [JsonProperty("request_id")] public string RequestId { get; set; }
        [JsonProperty("response")] public ControlResponseBody Response { get; set; }
    }

    public class ControlResponseBody
    {
        [JsonProperty("allowed")] public bool Allowed { get; set; }
        [JsonProperty("always")] public bool? Always { get; set; }
        [JsonProperty("session")] public bool? Session { get; set; }
        [JsonProperty("reason")] public string Reason { get; set; }
    }

    // =========================================================================
    // Incoming messages — Bun → IDE (match protocol.ts L164-446)
    // =========================================================================

    public class IncomingMessage
    {
        [JsonProperty("type")] public string Type { get; set; }
    }

    public class StatusMessage : IncomingMessage
    {
        [JsonProperty("status")] public string Status { get; set; }
    }

    public class AssistantMessage : IncomingMessage
    {
        [JsonProperty("message")] public JToken Message { get; set; }
        [JsonProperty("parent_tool_use_id")] public string ParentToolUseId { get; set; }
    }

    public class PartialAssistantMessage : IncomingMessage
    {
        [JsonProperty("message")] public JToken Message { get; set; }
        [JsonProperty("parent_tool_use_id")] public string ParentToolUseId { get; set; }
    }

    public class ResultMessage : IncomingMessage
    {
        [JsonProperty("subtype")] public string Subtype { get; set; } // "success" | "error"
        [JsonProperty("result")] public string Result { get; set; }
        [JsonProperty("error")] public string Error { get; set; }
    }

    public class ControlRequestMessage : IncomingMessage
    {
        [JsonProperty("request_id")] public string RequestId { get; set; }
        [JsonProperty("request")] public ControlRequestBody Request { get; set; }
    }

    public class ControlRequestBody
    {
        [JsonProperty("subtype")] public string Subtype { get; set; }
        [JsonProperty("tool_name")] public string ToolName { get; set; }
        [JsonProperty("tool_use_id")] public string ToolUseId { get; set; }
        [JsonProperty("input")] public JToken Input { get; set; }
        [JsonProperty("action_description")] public string ActionDescription { get; set; }
    }

    public class SessionMeta
    {
        [JsonProperty("sessionId")] public string SessionId { get; set; }
        [JsonProperty("title")] public string Title { get; set; }
        [JsonProperty("messageCount")] public int MessageCount { get; set; }
        [JsonProperty("timestamp")] public string Timestamp { get; set; }
        [JsonProperty("pinned")] public bool Pinned { get; set; }
    }

    public class SessionListMessage : IncomingMessage
    {
        [JsonProperty("sessions")] public List<SessionMeta> Sessions { get; set; }
    }

    public class ContextWindowMessage : IncomingMessage
    {
        [JsonProperty("context_window_size")] public int ContextWindowSize { get; set; }
        [JsonProperty("used_tokens")] public int UsedTokens { get; set; }
        [JsonProperty("used_percentage")] public double UsedPercentage { get; set; }
        [JsonProperty("remaining_percentage")] public double RemainingPercentage { get; set; }
        [JsonProperty("model")] public string Model { get; set; }
    }

    public class ErrorMessage : IncomingMessage
    {
        [JsonProperty("message")] public string Message { get; set; }
    }

    // =========================================================================
    // Helper: dispatch incoming messages by type field
    // =========================================================================

    public static class MessageTypes
    {
        // Outgoing (IDE → Bun)
        public const string User = "user";
        public const string IdeContext = "ide_context";
        public const string ControlResponse = "control_response";
        public const string Interrupt = "interrupt";
        public const string ListSessions = "list_sessions";
        public const string LoadSession = "load_session";
        public const string ResumeSession = "resume_session";
        public const string DeleteSession = "delete_session";
        public const string RenameSession = "rename_session";
        public const string NewSession = "new_session";
        public const string PinSession = "pin_session";
        public const string SetPermissionMode = "set_permission_mode";
        public const string SetThinkingMode = "set_thinking_mode";
        public const string SetModelProfile = "set_model_profile";
        public const string Compact = "compact";
        public const string SideQuestion = "side_question";
        public const string KillTask = "kill_task";
        public const string ListTasks = "list_tasks";
        public const string ListRewindPoints = "list_rewind_points";
        public const string ExecuteRewind = "execute_rewind";

        // Incoming (Bun → IDE)
        public const string Status = "status";
        public const string Assistant = "assistant";
        public const string PartialAssistant = "partial_assistant";
        public const string User_ = "user";
        public const string Result = "result";
        public const string Error = "error";
        public const string Log = "log";
        public const string ControlRequest = "control_request";
        public const string StreamEvent = "stream_event";
        public const string System = "system";
        public const string IdeContext_ = "ide_context";
        public const string SessionList = "session_list";
        public const string SessionLoaded = "session_loaded";
        public const string ContextWindow = "context_window";
        public const string TaskStarted = "task_started";
        public const string TaskProgress = "task_progress";
        public const string TaskCompleted = "task_completed";
        public const string TaskList = "task_list";
        public const string ToolProgress = "tool_progress";
        public const string FilePicked = "file_picked";
        public const string FillInput = "fill_input";
        public const string PermissionModeChanged = "permission_mode_changed";
        public const string PlanTasks = "plan_tasks";
        public const string RewindPoints = "rewind_points";
        public const string RewindCompleted = "rewind_completed";
        public const string SideQuestionResult = "side_question_result";
    }
}

File: extensions/vscode/src/protocol.ts
/**
 * Protocol types for IDE ↔ Claude Code communication over WebSocket.
 *
 * All messages are JSON objects with a `type` field. These types mirror
 * the corresponding interfaces in src/entrypoints/ideMode.ts.
 */

// ============================================================================
// Context types (IDE → Claude Code)
// ============================================================================

export interface IDEFileContext {
  path: string
  content: string
  language?: string
}

export interface IDESelection {
  path: string
  startLine: number
  startChar: number
  endLine: number
  endChar: number
  text: string
}

export interface IDEDiagnostic {
  path: string
  line: number
  column: number
  message: string
  severity: 'error' | 'warning' | 'info' | 'hint'
  source?: string
}

export interface FileTreeNode {
  name: string
  path: string
  isDir: boolean
  children?: FileTreeNode[]
}

// ============================================================================
// Messages FROM extension TO Claude Code (write to WebSocket)
// ============================================================================

export interface IDEUserMessage {
  type: 'user'
  message: { role: 'user'; content: string }
  parent_tool_use_id: string | null
  attachments?: { path: string; content?: string; language?: string; mediaType?: string; kind?: string; isDir?: boolean }[]
}

export interface IDEContextMessage {
  type: 'ide_context'
  files?: IDEFileContext[]
  selection?: IDESelection | null
  diagnostics?: IDEDiagnostic[]
}

export interface IDEControlResponseMessage {
  type: 'control_response'
  request_id: string
  response: {
    allowed: boolean
    always?: boolean
    session?: boolean
    reason?: string
    updatedInput?: Record<string, unknown>
  }
}

export interface IDEInterruptMessage {
  type: 'interrupt'
}

export interface IDESessionListRequest {
  type: 'list_sessions'
}

export interface IDESessionLoadRequest {
  type: 'load_session'
  session_id: string
}

export interface IDESessionResumeRequest {
  type: 'resume_session'
  session_id: string
}

export interface IDESessionDeleteRequest {
  type: 'delete_session'
  session_id: string
}

export interface IDESessionRenameRequest {
  type: 'rename_session'
  session_id: string
  title: string
}

export interface IDESessionNewRequest {
  type: 'new_session'
}

export interface IDESessionPinRequest {
  type: 'pin_session'
  session_id: string
  pinned: boolean
}

export interface IDETaskListRequest {
  type: 'list_tasks'
}

export interface IDEKillTaskRequest {
  type: 'kill_task'
  task_id: string
}

export interface IDESideQuestionMessage {
  type: 'side_question'
  question: string
  context_id: string
}

export interface IIDERewindPointsListRequest {
  type: 'list_rewind_points'
}

export interface IIDERewindExecuteRequest {
  type: 'execute_rewind'
  messageId: string
}

export interface IDESetPermissionModeRequest {
  type: 'set_permission_mode'
  mode: string
}

export type OutgoingMessage =
  | IDEUserMessage
  | IDEContextMessage
  | IDEControlResponseMessage
  | IDEInterruptMessage
  | IDESessionListRequest
  | IDESessionLoadRequest
  | IDESessionResumeRequest
  | IDESessionDeleteRequest
  | IDESessionRenameRequest
  | IDESessionNewRequest
  | IDESessionPinRequest
  | IDETaskListRequest
  | IDEKillTaskRequest
  | IDESetPermissionModeRequest
  | IDECompactRequestMessage
  | IDESideQuestionMessage
  | IIDERewindPointsListRequest
  | IIDERewindExecuteRequest

// ============================================================================
// Messages FROM Claude Code TO extension (read from WebSocket)
// ============================================================================

export interface IDEStatusMessage {
  type: 'status'
  status: string | null
}

export interface IDEAssistantMessage {
  type: 'assistant'
  message: Record<string, unknown>
  parent_tool_use_id: string | null
}

export interface IDEToolResultMessage {
  type: 'user'
  message: Record<string, unknown>
  parent_tool_use_id: string | null
}

export interface IDEResultMessage {
  type: 'result'
  subtype: 'success' | 'error'
  result?: string
  error?: string
}

export interface IDEErrorMessage {
  type: 'error'
  message: string
}

export interface IDELogMessage {
  type: 'log'
  message: string
}

export interface IDEConnectionStatusMessage {
  type: 'status'
  status: string | null
  code?: number
  reason?: string
}

export interface IDEControlRequestMessage {
  type: 'control_request'
  request_id: string
  request: {
    subtype: string
    tool_name: string
    tool_use_id: string
    input: Record<string, unknown>
    action_description: string
  }
}

export interface IDEStreamEventMessage {
  type: 'stream_event'
  event: unknown
  parent_tool_use_id: string | null
}

export interface IDEPartialAssistantMessage {
  type: 'partial_assistant'
  message: Record<string, unknown>
  parent_tool_use_id: string | null
}

export interface IDESystemMessage {
  type: 'system'
  subtype: string
  [key: string]: unknown
}

export interface IDEContextDisplayMessage {
  type: 'ide_context'
  files: string[]
  selection?: { path: string; lines: number; text: string } | null
  diagnostics?: { errors: number; warnings: number; info: number } | null
}

export interface IDEFillInputMessage {
  type: 'fill_input'
  text: string
  selection?: {
    filePath: string
    startLine: number
    endLine: number
  }
}

export interface IDEFilePickedMessage {
  type: 'file_picked'
  files: { path: string; content?: string; language?: string; isDir?: boolean }[]
}

export interface IDESessionMeta {
  sessionId: string
  title: string
  messageCount: number
  timestamp: string
  gitBranch?: string
  tag?: string
  pinned?: boolean
}

export interface IDESessionListResponse {
  type: 'session_list'
  sessions: IDESessionMeta[]
}

export interface IDESessionLoadedResponse {
  type: 'session_loaded'
  session_id: string
  messages: Array<{
    uuid?: string
    type: string
    message: Record<string, unknown>
    parent_tool_use_id?: string | null
    timestamp?: string
  }>
}

export interface IDESessionDeletedResponse {
  type: 'session_deleted'
  session_id: string
  was_current?: boolean
}

export interface IDESessionRenamedResponse {
  type: 'session_renamed'
  session_id: string
  title: string
}

export interface IDESessionCreatedResponse {
  type: 'session_created'
  session_id: string
}

export interface IDETaskStartedMessage {
  type: 'task_started'
  task_id: string
  description: string
  agent_type: string
}

export interface IDETaskProgressMessage {
  type: 'task_progress'
  task_id: string
  description: string
  status: string
  tool_uses: number
  total_tokens: number
}

export interface IDETaskCompletedMessage {
  type: 'task_completed'
  task_id: string
  description: string
  status: 'completed' | 'failed' | 'killed'
  error?: string
}

export interface IDETaskListMessage {
  type: 'task_list'
  tasks: Array<{
    task_id: string
    description: string
    agent_type: string
    status: string
  }>
}

export interface IDEPermissionModeChangedMessage {
  type: 'permission_mode_changed'
  mode: string
}

export interface IDEContextWindowMessage {
  type: 'context_window'
  context_window_size: number
  used_tokens: number
  used_percentage: number | null
  remaining_percentage: number | null
  model: string
}

export interface IDEPlanTasksMessage {
  type: 'plan_tasks'
  tasks: Array<{
    id: string
    subject: string
    description: string
    status: 'pending' | 'in_progress' | 'completed'
    owner?: string
    blockedBy: string[]
    blocks: string[]
    activeForm?: string
  }>
}

export interface IDESideQuestionResultMessage {
  type: 'side_question_result'
  context_id: string
  response?: string
  error?: string
}

export interface IDEClipboardMessage {
  type: 'clipboard'
  text: string
}

export interface IDEToolProgressMessage {
  type: 'tool_progress'
  data: {
    type: 'bash_progress'
    output: string
    fullOutput: string
    elapsedTimeSeconds: number
    totalLines: number
    totalBytes?: number
    taskId?: string
    timeoutMs?: number
  }
  tool_use_id: string
  parent_tool_use_id: string
}

export interface IDECompactRequestMessage {
  type: 'compact'
}

export interface IRewindPoint {
  messageId: string
  label: string
  timestamp: string
  filesChanged: number
  insertions: number
  deletions: number
}

export interface IRewindPointsMessage {
  type: 'rewind_points'
  points: IRewindPoint[]
}

export interface IRewindCompletedMessage {
  type: 'rewind_completed'
  messageId: string
  filesChanged: string[]
}

export type IncomingMessage =
  | IDEStatusMessage
  | IDEAssistantMessage
  | IDEToolResultMessage
  | IDEResultMessage
  | IDEErrorMessage
  | IDELogMessage
  | IDEConnectionStatusMessage
  | IDEControlRequestMessage
  | IDEStreamEventMessage
  | IDEPartialAssistantMessage
  | IDESystemMessage
  | IDEContextDisplayMessage
  | IDEFillInputMessage
  | IDEFilePickedMessage
  | IDESessionListResponse
  | IDESessionLoadedResponse
  | IDESessionDeletedResponse
  | IDESessionRenamedResponse
  | IDESessionCreatedResponse
  | IDETaskStartedMessage
  | IDETaskProgressMessage
  | IDETaskCompletedMessage
  | IDETaskListMessage
  | IDEPermissionModeChangedMessage
  | IDEContextWindowMessage
  | IDEPlanTasksMessage
  | IDEToolProgressMessage
  | IDESideQuestionResultMessage
  | IDEClipboardMessage
  | IRewindPointsMessage
  | IRewindCompletedMessage

// ============================================================================
// Type guard
// ============================================================================

export function isIncomingMessage(v: unknown): v is IncomingMessage {
  return typeof v === 'object' && v !== null && 'type' in v
}

/**
 * Parse a WebSocket message (always a JSON string) into a typed message.
 */
export function parseMessage(data: string): IncomingMessage | null {
  try {
    const parsed = JSON.parse(data)
    return isIncomingMessage(parsed) ? parsed : null
  } catch {
    return null
  }
}
